coldfront/core/department/utils/ldap.py

```diff
@@ -0,0 +1,118 @@
+    conn.search(DEPARTMENT_OU,
+                f'(&(objectClass=organizationalUnit)(ou={code}))',
+                attributes=['berkeleyEduOrgUnitHierarchyString'])
+    if not conn.entries:
+        return None
+    hierarchy = conn.entries[0].berkeleyEduOrgUnitHierarchyString.value
+    return hierarchy.split('-')[3]
+
+
+def get_department_name_from_code(code):
+    """Return the name of a department from its code (any level but ideally only L4).
+    Example: "IBIBI" -> "Integrative Biology"
+
+    Parameters:
+       code (str): department code
+    Returns:
+        (str) Name of department
+    """
+    conn = Connection(LDAP_URL, auto_bind=True)
+    conn.search(DEPARTMENT_OU,
+                f'(&(objectClass=organizationalUnit)(ou={code}))',
+                attributes=['description'])
+    return conn.entries[0].description.value
+
+
+def fetch_and_set_user_departments(user, userprofile, dry_run=False):
+    """Fetch a user's departments from LDAP and set them in the
+    database.
+
+    Parameters:
+        user (User): Django User object
+        userprofile (UserProfile): Coldfront UserProfile object
+        dry_run (bool): If True, don't actually make changes
+                        (default: False)
+    Returns:
+        None
+    """
+    if dry_run:
+        logger.info = print
+    user_entry = ldap_search_user(user.email, user.first_name, user.last_name)
+    ldap_departments = user_entry.departmentNumber.values if user_entry else []
+    for department_code in ldap_departments:
+        department_code = get_L4_code_from_department_code(department_code)
+        if department_code is None:
+            logger.warning(f'Could not find L4 code for department {department_code}')
+            continue
+        # If a Department doesn't exist, create it.
+        name = None if Department.objects.filter(code=department_code).exists()\
+            else get_department_name_from_code(department_code)
+        if dry_run:
+            department = Department(code=department_code, name=name)
+            userdepartment = UserDepartment(userprofile=userprofile,
+                                department=department, is_authoritative=True)
+            created = name is not None
+        else:
+            department, created = Department.objects.get_or_create(
+                                                    code=department_code,
+                                                    defaults={'name': name})
+        if created:
+            logger.info(f'Created Department {department}')
+
+        # Create a UserDepartment association, updating is_authoritative if
+        # needed.
+        if not UserDepartment.objects.filter(userprofile=userprofile,
+                                             department__code=department_code,
+                                             is_authoritative=True).exists():
+            if dry_run:
+                print(f'Created UserDepartment {userdepartment}')
```

Comment:
My IDE is warning me that `userdepartment` may be referenced before assignment. I think you can move line 97 down here.

Reply:
my vscode doesn't give that but I added it, afaik if dry_run is true then it's always defined on 97